src/camera.rs

```diff
@@ -86,18 +90,27 @@ impl Camera {
     ) -> Option<(Color, f32)> {
         let dist = center - self.transform.position;
         let a = ray.sq_mag();
+        // `a` will always be positive, let LLVM know so this can be optimized
+        unsafe {
+            std::intrinsics::assume(a >= 0.0);
+        }
+
         let b = ray.dot(dist);
         let c = dist.sq_mag() - r.powi(2);
 
-        let sqrt_term = (b.powi(2) - a * c).sqrt();
-        if !sqrt_term.is_finite() {
+        let sqrt_term_inner = b.powi(2) - a * c;
+        if (0.0 > sqrt_term_inner) || sqrt_term_inner.is_subnormal() {
```

Comment:
These parentheses shouldn't be necessary, but shouldn't the compiler have warned about that?
```suggestion
        if 0.0 > sqrt_term_inner || sqrt_term_inner.is_subnormal() {
```